Firmware/LowLevel/src/main.cpp

```diff
@@ -547,13 +561,81 @@ void onUIPacketReceived(const uint8_t *buffer, size_t size) {
     }
 }
 
-void sendConfigMessage(uint8_t pkt_type) {
-    struct ll_high_level_config ll_config;
-    ll_config.type = pkt_type;
-    ll_config.config_bitmask = config_bitmask;
-    ll_config.volume = 80;            // FIXME: Adapt once nv_config or improve-sound got merged
-    strncpy(ll_config.language, "en", 2); // FIXME: Adapt once nv_config or improve-sound got merged
-    sendMessage(&ll_config, sizeof(struct ll_high_level_config));
+void sendConfigMessage(const uint8_t pkt_type) {
+    size_t msg_size = sizeof(struct ll_high_level_config) + 3;  // + 1 type + 2 crc
+    uint8_t *msg = (uint8_t *)malloc(msg_size);
+    if (msg == NULL)
+        return;
+    *msg = pkt_type;
+    memcpy(msg + 1, &llhl_config, sizeof(struct ll_high_level_config));  // Copy our live config into the message, behind type
+    sendMessage(msg, msg_size);
+    free(msg);
+}
+
+/**
+ * @brief applyConfig applies those members who are not undefined/unknown.
+ * This function get called either when receiving a ll_high_level_config packet from HL, or during boot after read from LittleFS
+ * @param buffer
+ * @param size
+ */
+void applyConfig(const uint8_t *buffer, const size_t size) {
+    // This is a flexible length packet where the size may vary when ll_high_level_config struct get enhanced only on one side.
+    // If payload size is larger than our struct size, ensure that we only copy those we know of = our struct size.
+    // If payload size is smaller than our struct size, copy only the payload we got, but ensure that the unsent member(s) have reasonable defaults.
+    size_t payload_size = min(sizeof(ll_high_level_config), size - 2);  // exclude crc
+
+    // Use a temporary rcv_config for easier member access.
+    // If payload is smaller (older), struct already contains reasonable defaults.
+    ll_high_level_config rcv_config;
+    memcpy(&rcv_config, buffer, payload_size);  // Copy payload to temporary config
+
+    // Use a temporary new_config for save sanity adaption before copy to our live config.
+    // It's important to use a clean one with LL-default values, and not the current live llhl_config one,
+    // because we need to ensure that re-commented mower_config parameter get the LL-default value and not the last saved one!
+    ll_high_level_config new_config;
+
+    // Take over of payload values. This could also be made the other way around, which would save the copy of the "leading"
+    // values, but I felt it's more secure to do it this way (in special for further packet enlargements)
+
+    // HL always force these member
+    new_config.options = rcv_config.options;
+    strncpy(new_config.language, rcv_config.language, 2);
```

Comment:
> Why do these need to be handled differently? You could one bit to indicate "keep defaults" and the language could be {0,0} for undefined.
> 
It's reasoned in some discussion I've had in past during Improve-Sound PR.
That time, it was said that the language should be always given by HL and the same with DFP_IS_5V.
And during switchover from bitmask to bitfields I recognized that's also true for the two new bit- switches.
In addition I couldn't imagine any option, where LL could be leading for an option.

> Thinking more about it, I'm wondering whether it wouldn't be better to avoid the bitmask completely because it will make extensions harder, not easier. If we add another bit, it would have to be off unset by default, otherwise an older ROS version will always disable it because it will send a 0 in that place. Whereas adding another boolean at the end of the structure should now be no additional effort and you have all freedom to choose defaults.
> 
I also thought shortly about a possible new option in the bitfields options. But don't see an issue with the fact that it has to be integrated then, as default=false option :confused: 

Bit masks are pretty effective and and I really do like the more modern and compact implementation via bitfields :heart_eyes: 


> Or instead of a boolean, there should probably be a tri-value on/off/undefined enum (or 1/0/-1 int8) that can be sent by ROS, following what you did for all the other settings.

Hmm... tri-value enums in a larger bitfield :stuck_out_tongue_winking_eye: 
But really?! We don't have any tri-state-option now, nor I can imagine about one, as well as we've now 3 used options out of 8!
Adding additional bools or even tri-state-switches later, when required or to simple-minded (by the implementor) remain a future option :wink:

Reply:
> That time, it was said that the language should be always given by HL and the same with DFP_IS_5V.

I thought for a long time that everything is leading in HL... 😉 But where's the difference between these two settings and let's say the volume or the max pitch? You have already given a default language, so why shouldn't that be used unless explicitly overridden?

> But don't see an issue with the fact that it has to be integrated then, as default=false option 😕

You can always invert the meaning, sure. But let's say you want do enable background sounds by default. Then the setting would be no_background_sound=false... Double negation.

> We don't have any tri-state-option now, nor I can imagine about one

We will need that when we add boolean options. For the other data types, you can use -1 or 0xff as "undefined" value from the ROS side, but with a boolean, that will be difficult.

Reply:
> > That time, it was said that the language should be always given by HL and the same with DFP_IS_5V.
> 
> I thought for a long time that everything is leading in HL... 😉 But where's the difference between these two settings and let's say the volume or the max pitch? You have already given a default language, so why shouldn't that be used unless explicitly overridden?

I know, I used the "leading" phrase the first time related to this topic :innocent: ... it's probably miss leading :thinking: 

Volume could be controlled from both sides, and the user is free to decide if HL is forcing a specific volume or if it's changeable i.e. via CoverUI. Same as with language, it was also configurable from both sides in past, but got changed some months ago.
Both of them needs to be known during boot.
Similar to "max pitch", which should have a reasonable value (if enabled), even if HL is offline. But for "max pitch" we only have HL as controlling side at the moment. Very unlikely, but if one plans to integrate it i.e. into a CoverUI with Display, he probably need to change that.


> 
> > But don't see an issue with the fact that it has to be integrated then, as default=false option 😕
> 
> You can always invert the meaning, sure. But let's say you want do enable background sounds by default. Then the setting would be no_background_sound=false... Double negation.

Yes, that's clear to me, but sound also very theoretical to me. Well background sound would be uncritical because it could wait till it receives a config update from HL, but "yes" if there's a critical option in future, which need to be on by default, the implementer need to use a double negation label.

But imagine:
Isn't "IGNORE CHARGING CURRENT" in real a negated "ENABLE CHARGE CURRENT PROTECTION" ?

> 
> > We don't have any tri-state-option now, nor I can imagine about one
> 
> We will need that when we add boolean options. For the other data types, you can use -1 or 0xff as "undefined" value from the ROS side, but with a boolean, that will be difficult.

Okay, you're thinking about an option LL could have, but HL doesn't has a preference for.
Yip, for such a case we need a tri-state option. 
Will check to change the bitfield from boolean to a tri-state (enum) option.

Reply:
Done